Revert "Fix for NCE programming mode issues."

bobjacobsen opened the pull request:
Reverts JMRI/JMRI#837

The underlying problem with NCE programming needs to be fixed another way.

Fixing a broken NceProgrammerManager by adding hard-to-find code to the NceSystemConnectionMemo will just cause confusion in the future.

bobjacobsen commented:
(Will add problem fix to this branch shortly)

dheap commented:
I'll be interested to see how you manage to fix it another way.

The "hard to find code" was already in the NceSystemConnectionMemo.provides(Class) method. I just added it to the NceSystemConnectionMemo.get(Class) method as well.

It is also already in the CanSystemConnectionMemo.get(Class)  and the SprogSystemConnectionMemo.provides(Class) method. Are you going to remove it from these classes as well?

> The underlying problem with NCE programming needs to be fixed another way.
> Fixing a broken NceProgrammerManager by adding hard-to-find code to the NceSystemConnectionMemo will just cause confusion in the future.

rhwood commented on java/test/jmri/jmrix/nce/NceProgrammerManagerTest.java:
```diff
@@ -0,0 +1,77 @@
+        // this is checking the "as default ctor built" options, which might not be valid
+        Assert.assertTrue("provides global programmerManager", memo.provides(GlobalProgrammerManager.class));
+        Assert.assertNotNull("global ProgrammerManager exists", memo.get(GlobalProgrammerManager.class));
+        Assert.assertNotNull("global Programmer exists", ((GlobalProgrammerManager)memo.get(GlobalProgrammerManager.class)).getGlobalProgrammer());
```
Why is this cast here?

bobjacobsen replied:
> Why is this cast here?

Because without it, there's a compile error:
````
    [javac] Compiling 1 source file to /Users/jake/Documents/Trains/JMRI/projects/JMRI/java/classes
    [javac] /Users/jake/Documents/Trains/JMRI/projects/JMRI/java/test/jmri/jmrix/nce/NceProgrammerManagerTest.java:41: error: cannot find symbol
    [javac]         Assert.assertNull("no global Programmer exists", memo.get(GlobalProgrammerManager.class).getGlobalProgrammer());
    [javac]                                                                                                 ^
    [javac]   symbol:   method getGlobalProgrammer()
    [javac]   location: class Object
    [javac] 1 error
````

bobjacobsen commented:
Changes included:

- Fixing the NceProgrammerManager so that it properly handles the GlobalProgrammerManager interface:  isGlobalProgrammerAvailable() == false and getGlobalProgrammer() == null go together.
- Added JUnit tests that show desired behavior of NceSystemConnectionMemo and NceProgrammerManager
- NceSystemConnectionMemo returns a (now properly behaving) NceProgrammerManager always

While testing, also did a little cleanup:
- removed some $Revision$ markers
- made some logging more robust, including having java/src/jmri/jmrit/AbstractIdentify use it's given Programmer instead of looking up one of its own
- converted some "new" calls to Java 8 diamond notation (i.e. new List<> instead of new List<type>)
- made some jmrit.progsupport code consistent from class to class, as a step toward some (eventual) refactoring
- removed an odd call to InstanceManager that seemed to be hoping that something good comes from failure

And added a new feature:
- DecoderPro (RosterFrame) now disables the Ident (loco identify) button if there's no service mode programmer.

bobjacobsen commented:
> It is also already in the CanSystemConnectionMemo.get(Class)  and the SprogSystemConnectionMemo.provides(Class) method. Are you going to remove it from these classes as well?

Thanks for pointing that out. yes, while I'm spending my Saturday cleaning up code, I might as well do that.  But I'll put it in a different PR.

bobjacobsen commented:
@dheap has offered to test with NCE hardware (Thanks!), so this is staying labeled as WIP for now.

KenC57 commented:
Reading through the code I'm not sure when there is or isn't a programmer.
From my view of NCE systems, all have a addressed programmer. But only the
PowerCab and Twin have programming track modes on the USB. I didn't know the
Twin supported it, but just downloaded the manual and (with a ProCab
throttle) you get an interface to do the programming. It even shows using
the Auto-SW to isolate the layout from the programming track.

 

I don't think any of the logic knows that right now.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

dheap commented:
Have soon hit problems with hardware testing:

First test is with single connection to SB5 (has no service mode, only an
ops mode).

- [x] Identify is disabled (as it should be).
- [ ] Create New Loco pane shows Programming Mode with only ... in
pulldown box.
- [ ] Create New Loco pane allows you to manually select a decoder and
"Open Comprehensive Programmer". But it opens in Service Mode (an invalid
mode) and there are no Write buttons. It should open in Ops Mode and allow
you to Write (but not Read).
- [ ] Selecting an existing roster entry and opening in POM mode generates
an Exception in the log (below), but does allow you to go ahead and write
anyway.

I've checked back and the first two were not addressed by my #837 either,
but it did remove the not-wanted Programming track and Programming Mode
options from Roster Frame and open roster entries.

Due to the time involved in lots of hardware testing, I'll suspend further
testing until we resolve the above...

>@dheap has offered to test with NCE hardware (Thanks!), so this is
>staying labeled as WIP for now.

2016-02-21 15:23:42,529 nce.NceProgrammer                     WARN  -
NceProgrammer cannot function, shouldn't have been created
[AWT-EventQueue-0]
java.lang.Exception: traceback
	at jmri.jmrix.nce.NceProgrammer.<init>(NceProgrammer.java:31)
	at 
jmri.jmrix.nce.NceOpsModeProgrammer.<init>(NceOpsModeProgrammer.java:32)
	at 
jmri.jmrix.nce.NceProgrammerManager.getAddressedProgrammer(NceProgrammerMan
ager.java:69)
	at 
jmri.jmrit.roster.swing.RosterFrame.startProgrammer(RosterFrame.java:1425)
	at 
jmri.jmrit.roster.swing.RosterFrame.lambda$bottomRight$45(RosterFrame.java:
296)
	at 
javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at 
javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348
)
	at 
javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:
402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at 
javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListene
r.java:252)
	at java.awt.Component.processMouseEvent(Component.java:6535)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6300)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4891)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4713)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
	at java.awt.Container.dispatchEventImpl(Container.java:2280)
	at java.awt.Window.dispatchEventImpl(Window.java:2750)
	at java.awt.Component.dispatchEvent(Component.java:4713)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at 
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivile
ge(ProtectionDomain.java:76)
	at 
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivile
ge(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at 
java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivile
ge(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at 
java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.jav
a:201)
	at 
java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:1
16)
	at 
java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.jav
a:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

bobjacobsen commented:
````
Reading through the code I'm not sure when there is or isn't a programmer.
From my view of NCE systems, all have a addressed programmer. But only the
PowerCab and Twin have programming track modes on the USB. I didn't know the
Twin supported it, but just downloaded the manual and (with a ProCab
throttle) you get an interface to do the programming. It even shows using
the Auto-SW to isolate the layout from the programming track.

I don't think any of the logic knows that right now.
````

The place to address that is in the NceProgrammer class.  Its isGlobalProgrammerAvailable() and getGlobalProgrammer() methods should know about when the NceProgrammer can successfully perform programming operations and when it can't.

bobjacobsen commented:
@dheap:
````
First test is with single connection to SB5 (has no service mode, only an
ops mode).

- [x] Identify is disabled (as it should be).
- [  ] Create New Loco pane shows Programming Mode with only ... in
pulldown box.
- [  ] Create New Loco pane allows you to manually select a decoder and
"Open Comprehensive Programmer". But it opens in Service Mode (an invalid
mode) and there are no Write buttons. It should open in Ops Mode and allow
you to Write (but not Read).
- [  ] Selecting an existing roster entry and opening in POM mode generates
an Exception in the log (below), but does allow you to go ahead and write
anyway.
````

Thanks for doing the testing. Since it's back to working as well as v4.3.2, I'm going to merge this.  When there's time available, I can perhaps take a look at those other issues, but it's not code I'm familiar with.

dheap commented:
We've already addressed that in the mentioned code long ago, except for the new information Ken has come across about the Twin. We didn't think it had a GlobalProgrammer but it now appears that it does.

Probably not a good idea to try changing NceProgrammer while we are sorting out the mode issues. It can be addressed once any PR(s) are merged.

>  @bobjacobsen The place to address that is in the NceProgrammer class. Its isGlobalProgrammerAvailable() and getGlobalProgrammer() methods should know about when the NceProgrammer can successfully perform programming operations and when it can't.

dheap commented:
My #837 may have been stylistically undesirable but worked better than both V4.3.2 and this one (#848). It also matched existing practice in a several programmers.

> @bobjacobsen Thanks for doing the testing. Since it's back to working as well as v4.3.2, I'm going to merge this. When there's time available, I can perhaps take a look at those other issues, but it's not code I'm familiar with.

dheap commented:
My #837 was also thoroughly tested on all hardware and in all cases worked better than V4.3.2, with no Exceptions being thrown.

bobjacobsen commented:
@dheap:
> My #837 may have been stylistically undesirable but worked better than both V4.3.2 and this one (#848). It also matched existing practice in a several programmers.

In the "code wins" context, these kinds of issues are always tricky.

That change was _wrong_, not just stylistically undesirable.  You still don't seem to understand the difference between a ProgrammerManager and a Programmer. 

It's true that the same error is also elsewhere. But that doesn't mean that more errors should be made.

It's also true that there were other errors that were uncovered in all this work.  That also happens.

Because code wins, I've spent a bunch of time cleaning up other people's code; most of a Saturday.  Such is life, that's how things progress around here. 

If you want to contribute to this, move on to fixing the next things.

bobjacobsen commented:
````
- [  ] Create New Loco pane allows you to manually select a decoder and
"Open Comprehensive Programmer". But it opens in Service Mode (an invalid
mode) and there are no Write buttons. It should open in Ops Mode and allow
you to Write (but not Read).
```

This has been there for a long time. #851 is a new issue that describes what I've learned about what's happening.

dheap commented:
We'll have to agree to disagree on this one. But in this case I still believe it is a bad decision; I don't like code that results in an Exception that the user will see in the System Console and I don't accept the accusation that I don't understand the difference between a ProgrammerManager and a Programmer.
> @dheap:
> 
> My #837 may have been stylistically undesirable but worked better than both V4.3.2 and this one (#848). It also matched existing practice in a several programmers.
> 
> In the "code wins" context, these kinds of issues are always tricky.
> 
> That change was wrong, not just stylistically undesirable. You still don't seem to understand the difference between a ProgrammerManager and a Programmer.

bobjacobsen commented:
> I don't like code that results in an Exception that the user will see in the System Console 

Right.  That should be fixed. 

You do realize that it's a diagnostic traceback exception, right?

```
nce.NceProgrammer                     WARN  -
NceProgrammer cannot function, shouldn't have been created
[AWT-EventQueue-0]
java.lang.Exception: traceback
	at jmri.jmrix.nce.NceProgrammer.<init>(NceProgrammer.java:31)
```

KenC57 commented:
And I think the right way things should work now is any change of what
system supports what should only impact the isGlobalProgrammerAvailable()
and that the getGlobalProgrammer() checks that to know if it should use a
null or not. Since there is only one programmer, that is what I think works.
Let me know if I'm still off base.

 

But in the end of the day I want the details of figuring out what can or
can't be done with the particular system to be in as singular bit of code as
possible. In the past that wasn't always the case and now when we find (or
need to add) new things tracking down all the relevant bits is a pain.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

dheap commented:
What system supports what also impacts isAddressedModePossible(). We currently always return true from isAddressedModePossible() and we should not. NCE USB connected to a Power Pro (current version) does not support  programming and we need to check for that, and make it easy to change in the future.

My understanding of what Bob is saying is that we need to move system-specific conditionals and programmer reservations out of NceSystemConnectionMemo and into NceProgrammerManager. 

That sounds like a good idea but we will also need to be able to get an AddressedProgrammer class instance or a null return if unavailable for that specific system/connection combination i.e. isAddressedModePossible(). All current NceOpsModeProgrammer constructors require a loco address. 

> @KenC57
> And I think the right way things should work now is any change of what
> system supports what should only impact the isGlobalProgrammerAvailable()
> and that the getGlobalProgrammer() checks that to know if it should use a
> null or not. Since there is only one programmer, that is what I think works.
> Let me know if I'm still off base.

KenC57 commented:
The first paragraph of the NCE-USB manual says that ops mode is supported
for any system. I'm going to check further details on this in the documents
first. Then I'll setup a test on my system.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

KenC57 commented:
PS: later in the same document, it says there is a PowerPro command station
bug, so it won't support ops mode at this time. But I'll check and confirm
if that is true of all PowerPro or just which versions of the firmware for
the command station.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

KenC57 commented:
I'm building a list of spots where things like getCmdGroups() and such are
not right. Example would be UsbDriverAdapter():147, it is saying that the
ops commands are supported for the PowerPro. Technically right for the USB
side, but since that part of the code also knows things like firmware
versions in the PowerPro, we'd know that end doesn't support it. So I'm
inclined to say this should also mark that ops isn't supported. Later
if/when the PowerPro firmware is updated so it works, we add the case that
includes that feature.

 

Now looking at how these interact I'm also thinking that the right thing in
a number of other places would be just looking at the getCmdGroups() to see
if some operation/feature is supported or not instead of looking for the
type of interface. The granularity of the getCmdGroups() is finer than the
USB types I think.

 

Either way, I'm going to watch and wait for things to settle down first.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

dheap commented:
We are currently using 0x93-0x96 which are being trapped by JMRI code as unsupported. This is backed by the list of supported commands on page 5 of the V7.3.x manual I have.

There is a new  USB-only 0xAE command for Ops Mode Loco CV programming documented on page 10, but page 5 states that the Power Pro does not support it due to firmware bug.

We need to do some non-JMRI direct communications tests and verify what is going on.

> @KenC57
> The first paragraph of the NCE-USB manual says that ops mode is supported
> for any system. I'm going to check further details on this in the documents
> first. Then I'll setup a test on my system.

KenC57 commented:
Ok, the range 0x93->0x96 are the are the commands to queue a packet into the
Temp queue of the command station. These are used to send a set of bytes to
the track. So on some systems, this is how the addressed mode programmer
works. Now commands 0xA3->0xA5 look real similar. Only difference is these
say 'track' queue instead of 'temp_q', so whatever that difference means.
But both of those command sets are only on the serial mode.

 

So a question is how a PowerPro (serial) goes about creating the ops
programming packets?? I can't picture the cab doing it (creating a packet
and sending that to the command station via one of the queue commands), they
just aren't that smart (for everything else).

 

The right fix as I see it has the isAddressed. needs to say no for the
serial command station over the USB, and I don't think it does that right
now. So it is like I wrote the code to the first part of the document
without seeing the tail end of the document. That is until the day that a
new command station firmware comes out that knows how to respond to the 0xAE
command. Then for that version, we then allow ops programming.

 

So Dave, does this sound like what you are thinking?? I'm hoping Monday to
get confirmation of what versions should be doing what or not.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

dheap commented:
See below:

> @KenC57
> Ok, the range 0x93->0x96 are the are the commands to queue a packet into the
> Temp queue of the command station. These are used to send a set of bytes to
> the track. So on some systems, this is how the addressed mode programmer
> works. Now commands 0xA3->0xA5 look real similar. Only difference is these
> say 'track' queue instead of 'temp_q', so whatever that difference means.
> But both of those command sets are only on the serial mode.

It's a bit more complex than that. The fine print for the 0xA3->0xA5 track queue commands says (replaces any packet with the same address if it exists). I suspect these commands are for uses like overriding the repeating speed step commands for that loco. I would interpret the track queue as possibly being the entries in this rotating loop. (I'm trying to read between the lines.)

On the other hand the 0x93->0x96 TEMP_Q commands have a repeat count as the first byte and seem intended for ops mode packets needing a finite repeat count.

What is of more interest is the new 0xAE ops program loco CV command (marked ***USB COMMAND ONLY***).

Maybe that is what we should be using this new command with the V7.x.x boards and the V1.65x Power Cab/SB(3,5)X/Twin systems.

But for the Power Pro "2012 version"? 0xAE is marked as "not supported due to bug in command station software".

> So a question is how a PowerPro (serial) goes about creating the ops
> programming packets?? I can't picture the cab doing it (creating a packet
> and sending that to the command station via one of the queue commands), they
> just aren't that smart (for everything else).

The Pro Cab is basically a dumb terminal with a keyboard that sends keystrokes and an LCD display that is loaded in quadrants from the command station. Despite appearances, it doesn't even do a local echo (try watching the screen at the limits of radio reception, but be prepared to walk a hundred metres or more). All processing is done in the command station.

> The right fix as I see it has the isAddressed. needs to say no for the
> serial command station over the USB,

Correct.

> and I don't think it does that right
> now.

Correct.

> So it is like I wrote the code to the first part of the document
> without seeing the tail end of the document. That is until the day that a
> new command station firmware comes out that knows how to respond to the 0xAE
> command. Then for that version, we then allow ops programming.

Yes, as far as Power Pro is concerned.
> 
> So Dave, does this sound like what you are thinking?? I'm hoping Monday to
> get confirmation of what versions should be doing what or not.

I think we have a few cleanup tasks:
- Fix isAddressedModePossible() to return false for the Power Pro/USB combination.
- Make sure an addressed programmer is not loaded in that case.
- Consider replacing the 0x93->0x96 TEMP_Q commands with the new 0xAE ops program loco CV command on compatible USB system/version combinations.
- Address Bob's concerns that too much processing is being done in the NceSystemConnectionMemo commands rather than the NceProgrammerManager class.
- ...

Dave

KenC57 commented:
Dave,

 

Ok, I just looked at the NceOpsModeProgrammer (I didn't recall that it
wasn't just part of NceProgrammer) and I see ugly. It is using the command
on the USB but the NMRA packet for the serial stations. But the logic
between them isn't right, given what we've learned so far. But the NRMA
packet method is doing the 0x93->0x96, but it has a poor way to come up with
the command and if ever given a wrong size for the packet, you get the wrong
command.

 

-Ken Cameron, Member JMRI Dev Team

www.jmri.org

www.fingerlakeslivesteamers.org

www.cnymod.com

www.syracusemodelrr.org

bobjacobsen commented:
````
I think we have a few cleanup tasks:
...
- Address Bob's concerns that too much processing is being done in the NceSystemConnectionMemo
commands rather than the NceProgrammerManager class.
````
I think that's OK now.  The NceSystemConnectionMemo is providing a NceProgrammerManager when it should.  (Unless you decide to have more than implementation class for NCE ProgrammerManager implementation,) I think that's all set.

```
Ok, I just looked at the NceOpsModeProgrammer (I didn't recall that it
wasn't just part of NceProgrammer) and I see ugly. It is using the command
on the USB but the NMRA packet for the serial stations. But the logic
between them isn't right, given what we've learned so far. 
```
Code that tries to handle too many cases can get ugly.  Another approach is to create a number of separate Programmer classes, each handling only one kind of hardware. Each Programmer implementation is simpler, because it only handles one set of interactions.  I don't know what the right place to subdivide is, but it might be one for some USB, another for the rest of the USB choices, another for serial, etc. Then the ProgrammerManager, when asked for a Programmer, creates and returns one of the right type.

dheap commented:
We need to differentiate between the programmers we are getting.
- When NceSystemConnectionMemo asks for a a Global Programmer, we get a Global Programmer.
- When NceSystemConnectionMemo asks for a an Addressed Programmer, we use exactly the same code to get a Global Programmer, not an Addressed Programmer, but we then use it as an Addressed Programmer.

That is not good and is most likely the cause of the issue you were flagging with an exception.
> I think we have a few cleanup tasks:
> ...
> - Address Bob's concerns that too much processing is being done in the NceSystemConnectionMemo
> commands rather than the NceProgrammerManager class.
> I think that's OK now. The NceSystemConnectionMemo is providing a NceProgrammerManager when it should. (Unless you decide to have more than implementation class for NCE ProgrammerManager implementation,) I think that's all set.

dheap commented:
Sorry, I now realise my comments were applicable to the pre #848 state of NceSystemConnectionMemo and other classes and are no longer applicable.

We now get an NceProgrammerManager object, not a GlobalProgrammerManager object. Furthermore the code in RosterFrame has been reworked to check isAvailable... for each requested mode. (The availability will vary greatly, depending on the connection type combination; NCE USB vs serial port, system type, jumper configuration, firmware versions of both system and NCE USB. Sometimes one or other will be available, sometimes none and sometimes both. Such is the NCE environment.)

I was caught in somewhat of a time warp, so my earlier comment was inapplicable.

Dave

> We need to differentiate between the programmers we are getting.
> - When NceSystemConnectionMemo asks for a a Global Programmer, we get a Global Programmer.
> - When NceSystemConnectionMemo asks for a an Addressed Programmer, we use exactly the same code to get a Global Programmer, not an Addressed Programmer, but we then use it as an Addressed Programmer.
> 
> That is not good and is most likely the cause of the issue you were flagging with an exception.
